Add ItemActions

MrHell228 opened the pull request:
**SpongeAPI** | [Sponge](https://github.com/SpongePowered/Sponge/pull/4195)

Exposes `ConsumeEffect`s, adds`Keys.CONSUME_EFFECTS` and `Keys.DEATH_PROTECTION_EFFECTS` to apply them to `ItemStack`.

Currently `Keys.APPLICABLE_POTION_EFFECTS` does nothing and its `WeithedTable` approach doesn't make much sense anymore so I think it would be better to just delete it.

MrHell228 committed:
delete Keys.APPLICABLE_POTION_EFFECTS

MrHell228 committed:
make PotionEffectType Taggable

MrHell228 committed:
Add ConsumeEffects

MrHell228 committed:
fix some registries having not defaulted type

MrHell228 committed:
add Keys.DEATH_PROTECTION_EFFECTS

MrHell228 committed:
Merge branch 'api-14' into api-14-consume-effects

MrHell228 commented:
Actually, ConsumeEffect might be not the best name because DEATH_PROTECTION component also uses them. Can't think of a different name though

aromaa commented:
More fitting names could be `ItemActionEffect` or `EntityActionEffect`. I'm not really sure would we like to actually generalize this and then Mojang throws a curveball at us but food for thought.

MrHell228 committed:
rename ConsumeEffect -> ItemActionEffect

MrHell228 committed:
remove world arg for ItemActionEffect#apply

MrHell228 committed:
Merge branch 'api-14' into api-14-consume-effects

MrHell228 committed:
update ItemActionEffect docs

MrHell228 commented:
`ItemActionEffect` sounds fine, changed. Also removed `World` arg for `ItemActionEffect#apply` since it's only used for sounds and not really excepted to differ from entity world.
I'm not sure what you mean by not generalizing this. If it's about having like `Keys<List<Double>>` and `Key<List<SoundType>>` for teleports and sounds, then there is still no guarantee that mojang won't make it, for example, single teleport and single sound effect at some point.
So I guess nothing can be considered as a totally right choice for future here.

Yeregorix commented on src/main/java/org/spongepowered/api/data/type/ConsumeEffectTypes.java:
```diff
@@ -0,0 +1,63 @@
+ */
+@SuppressWarnings("unused")
+@RegistryScopes(scopes = RegistryScope.GAME)
+public final class ConsumeEffectTypes {
```
Should be `ItemActionEffectTypes`.

Yeregorix commented on src/main/java/org/spongepowered/api/data/Keys.java:
```diff
@@ -398,11 +399,6 @@ public final class Keys {
-    /**
-     * The set of {@link PotionEffect}s applied on use of an {@link ItemStack}.
-     */
-    public static final Key<WeightedCollectionValue<PotionEffect>> APPLICABLE_POTION_EFFECTS = Keys.weightedKey(ResourceKey.sponge("applicable_potion_effects"), PotionEffect.class);
```
Because API-14 have been released, this would be a breaking change. Rebase the PR to API-15 or keep this key for now and remove it in later versions.

MrHell228 committed:
rename to ItemAction

Yeregorix commented:
`ItemActionEffect` sounds a bit redundant to me. Wouldn't `ItemAction` or `ItemEffect` be enough? I have a slight preference for `ItemAction` as it is easier to distinguish from `PotionEffect`.

MrHell228 commented:
I don't mind renaming it to `ItemAction`. I guess it does reduce the understanding of the class purpose from the name but I believe that's what docs are for

Yeregorix commented:
Well, that's just my opinion, I'm not 100% sure

MrHell228 commented:
I prefer shorter names overall though here I'm not 100% sure as well :/

MrHell228 committed:
Revert "delete Keys.APPLICABLE_POTION_EFFECTS"

This reverts commit 5b7dab89e89b5f6532c77afc76262fe574f6b3c6.

MrHell228 committed:
Merge branch 'api-14' into api-14-consume-effects

MrHell228 commented:
Renamed to `ItemAction` (new keys got renamed as well) and returned deleted key (it had no implementation anyway)

Yeregorix commented:
Merged by 34622032e7bda623d9eff36cd00e7d1503cb2af6
